source/tools/CustomHostCpuManager.h

```diff
@@ -0,0 +1,477 @@
+                 * or to force the bandwidth update when the last measured
+                 * workload is above/below a given threshold.
+                 */
+                mBWUpdateCounter++;
+                if (mBWUpdateCounter >= mBWUpdatePeriod) {
+                    mBWUpdateCounter = 0;
+                    setApplicationLoad(getCurrentWorkUnits(), getMaxWorkUnits());
+                }
+            } else {
+                unboostBandwidth();
+            }
+
+            HostTools::sleepUntilNanoTime(wakeupTime_ns);
+        }
+
+        /* Read starting time of the callback */
+        getDeadlineAttribute(mDLStart);
+        mStartingTime_ns = HostTools::getNanoTime();
+    }
+
+    /**
+     * This is called when the application determines its workload for the
+     * callback.
+     *
+     * It calculates the runtime parameter to be assigned to the SCHED_DEADLINE
+     * Linux scheduling class, which adapts the CPU frequency accordingly.
+     *
+     * @param currentWorkUnits the current work units requested by the
+     * application, representing the computational effort that is going to be
+     * requested in the future computations. The meaning of this value may be
+     * different for different applications, e.g. it can be the number of notes
+     * played or the number of filters applied to a given signal.
+     * @param maxWorkUnits (not used) the maximum number of work units the
+     * application is planning to request.
+     */
+    void setApplicationLoad(int32_t currentWorkUnits,
+                            int32_t maxWorkUnits) override {
+        /*
+         * Perform an update only when the number of work units is changed, or
+         * if the mBWUpdateCounter is 0 (periodic update).
+         */
+        if (currentWorkUnits != getCurrentWorkUnits() ||
+                maxWorkUnits != getMaxWorkUnits() ||
+                mBWUpdateCounter == 0) {
+            int64_t expectedRuntime_ns;
+            double bandwidth;
+
+            /*
+             * mBWUpdateCounter is reset, to prevent a close bandwith updates
+             * due to the timer, after a workUnit change. This allows the
+             * statistics for the new workUnits to stabilize.
+             */
+            if (mBWUpdateCounter != 0)
+                mBWUpdateCounter = 0;
+
+            expectedRuntime_ns = mCpuTiming.getWorkUnitRuntime(currentWorkUnits);
+
+            /*
+             * When we read a 0 runtime for != 0 workUnits, it means that
+             * something bad is going on, so, set the maximum runtime.
+             * This may happen when the linear regression fails.
+             */
+            if (expectedRuntime_ns == 0 && currentWorkUnits != 0)
+                bandwidth = mMaxBW;
+            else
+                bandwidth = static_cast<double>(expectedRuntime_ns) / mPeriod_ns;
+
+            /* Add some offset to the computed bandwidth */
+            bandwidth *= 1.1;
```

Comment:
Explain these arbitrary values. Use constants.
Is the offset to adjust for application overhead?

Reply:
Solved in next push